File: benchmarks/file_io/csv_load.py
"""csv_load.py -- large multi-column CSV load benchmark, Python side.

Same shape as csv_load.qu (3,000,000 rows x 6 numeric columns), own RNG
stream (not bit-identical values, same statistics) -- pandas.read_csv is
the thing being timed.

Run: python benchmarks/file_io/csv_load.py
"""
import time
import numpy as np
import pandas as pd

N = 3_000_000
path = "benchmarks/file_io/py_wide_data.csv"

rng = np.random.default_rng(3)
df_out = pd.DataFrame({
    "c1": rng.standard_normal(N),
    "c2": rng.standard_normal(N),
    "c3": rng.standard_normal(N),
    "c4": rng.standard_normal(N),
    "c5": rng.standard_normal(N),
    "c6": rng.standard_normal(N),
})
df_out.to_csv(path, index=False)

t0 = time.perf_counter()
df = pd.read_csv(path)
t_load = time.perf_counter() - t0
print(f"pandas.read_csv ({N} rows x 6 cols) : {t_load:.4f} s")

print(f"mean(c1) = {df['c1'].mean():.6f}")
print(f"mean(c6) = {df['c6'].mean():.6f}")
print(f"nrow = {len(df)}  ncol = {df.shape[1]}")

File: benchmarks/file_io/csv_load.m
% csv_load.m -- large multi-column CSV load benchmark, MATLAB side.
%
% Same shape as csv_load.qu/.py (3,000,000 rows x 6 numeric columns).
% Uses readtable, per the task's ask for MATLAB's own idiomatic table
% loader. Matty has no readtable (checked builtins.py: only csvread/
% csvwrite) -- see csv_load_matty.m for the csvread variant run against
% Matty instead, and the README for that incompatibility noted explicitly.
%
% Run (cd into this script's own directory first, same reasoning as
% binary_doubles.m):
%   matlab -batch "cd('benchmarks/file_io'); csv_load"

N = 3000000;
bin_path = 'matlab_wide_data.csv';

rng(3);
T = table(randn(N,1), randn(N,1), randn(N,1), randn(N,1), randn(N,1), randn(N,1), ...
    'VariableNames', {'c1','c2','c3','c4','c5','c6'});
writetable(T, bin_path);

tic;
df = readtable(bin_path);
t_load = toc;
fprintf('readtable (%d rows x 6 cols) : %.4f s\n', N, t_load);

fprintf('mean(c1) = %.6f\n', mean(df.c1));
fprintf('mean(c6) = %.6f\n', mean(df.c6));
fprintf('nrow = %d  ncol = %d\n', height(df), width(df));
